multinet/auth/google.py

```diff
@@ -0,0 +1,159 @@
+"""Handling of Google Authorization."""
+import requests
+import base64
+import json
+
+from os import getenv
+from flask import Flask, redirect, request, make_response, url_for
+from werkzeug.wrappers import Response as ResponseWrapper
+from flask.blueprints import Blueprint
+from authlib.integrations.flask_client import OAuth
+
+from webargs.flaskparser import use_kwargs
+from webargs import fields
+
+from multinet.user import (
+    load_user_from_cookie,
+    load_user,
+    save_user,
+    get_user_cookie,
+    set_user_cookie,
+    register_user,
+    filter_user_info,
+    filter_document_meta,
+)
+
+from multinet.auth.types import GoogleUserInfo, User
+
+from typing import Dict
+
+
+CLIENT_ID = getenv("GOOGLE_CLIENT_ID")
+CLIENT_SECRET = getenv("GOOGLE_CLIENT_SECRET")
+
+GOOGLE_BASE_API = "https://www.googleapis.com/"
+GOOGLE_USER_INFO_URL = "oauth2/v3/userinfo"
+
+MULTINET_COOKIE = "multinet-token"
+
+bp = Blueprint("google", "google")
+oauth = OAuth()
+
+states_to_return_urls = {}
+
+
+def parse_id_token(token: str) -> GoogleUserInfo:
+    """Parse the base64 encoded id token."""
+    parts = token.split(".")
+    if len(parts) != 3:
+        # Should probably throw something here
+        pass
+
+    payload = parts[1]
+    padded = payload + ("=" * (4 - len(payload) % 4))
+    decoded = base64.b64decode(padded)
+    return json.loads(decoded)
+
+
+def ensure_external_url(url: str) -> str:
+    """Ensure a url is prefixed with a protocol (http or https)."""
+    return_url = url
+
+    if "http://" not in url and "https://" not in url:
+        return_url = f"http://{url}"
+
+    return return_url
+
+
+def google_oauth2_info() -> Dict:
+    """Return Google's spec for their OAuth endpoints."""
+    resp = requests.get("https://accounts.google.com/.well-known/openid-configuration")
+    info = resp.json()
+    return info
+
+
+def init_oauth(app: Flask) -> None:
+    """Initialize the OAuth integration."""
+    oauth.init_app(app)
+    info = google_oauth2_info()
+    oauth.register(
+        name="google",
+        client_id=CLIENT_ID,
+        client_secret=CLIENT_SECRET,
+        access_token_url=info["token_endpoint"],
+        authorize_url=info["authorization_endpoint"],
+        api_base_url=GOOGLE_BASE_API,
+        client_kwargs={"scope": "profile email openid"},
```

Comment:
>The scope parameter must begin with the openid value and then include the profile value, the email value, or both.
This is from the [docs](https://developers.google.com/identity/protocols/oauth2/openid-connect#scope-param)

Reply:
Also, do we need profile and email? I assume email just gives email and profile gives name, picture, etc.?

Reply:
> This is from the [docs](https://developers.google.com/identity/protocols/oauth2/openid-connect#scope-param). We might want to just change the order here to have parity with the spec.

Good catch, I'll do that and make sure it behaves like we expect.

> Also, do we need profile and email? I assume yes, and that email just gives email and profile gives name, picture, etc.?

Yes, I believe we'd need both of these.